[enhancement]: Custom resolver nameservers

hoxxep opened the pull request:
Add support for configuring custom nameservers in the configs.

Simplified resolver config example:
```toml
[resolver]
type = "custom"

[[resolver.nameservers]]
ip = "1.1.1.1"
port = 53
protocol = "udp"
```

williamdes reviewed (commented):
Thank you!

This will solve bugs like https://github.com/stalwartlabs/mail-server/issues/126

hoxxep commented:
Glad it's useful for more than myself! We can make the tests more thorough later if https://github.com/hickory-dns/hickory-dns/pull/2093 gets merged and we can inspect the resolver config.

stappersg commented:
> We can make the tests more thorough later if [hickory-dns/hickory-dns#2093](https://github.com/hickory-dns/hickory-dns/pull/2093) gets merged

It got merged.  `\o/` Hurray!

stappersg commented:
Please an _end of line_ to the last line of `tests/src/smtp/outbound/ip_lookup.rs`

![image](https://github.com/stalwartlabs/mail-server/assets/35434094/78fd90ff-491d-41f5-a2e9-6d62755aebee)

stappersg commented:
Thing I could not read in the code:   _Are `resolver` and `resolver.nameservers` custom resolver **optional**?_

hoxxep commented:
I'll wait until hickory release a v0.24.1 or equivalent with the change, and then will make a separate PR with the extra line to check the server in the tests.

Fixed the newline, and I've added to this MR a `.editorconfig` which a number of IDEs should respect by default to include trailing newlines.

stappersg commented:
On Mon, Nov 20, 2023 at 12:16:29PM -0800, Liam Gray wrote:
> I'll wait until hickory release a v0.24.1 or equivalent with the
> change, and then will make a separate PR with the extra line to check
> the server in the tests.

Fine with me.

 
> Fixed the newline,

Good, thanks.


> and I've added to this MR a `.editorconfig`

Not so good, it should go in a separate merge request.


Groeten
Geert Stappers
-- 
Silence is hard to parse

hoxxep commented:
@stappersg `[[resolver.nameservers]]` should be optional as we can instead specify default resolvers which don't require manual settings (eg. cloudflare, google, system). When specifying a custom resolver, the `[[resolver.nameservers]]` field is required, and `ip`, `port`, and `protocol` fields are all required. Potentially we could specify defaults.

Specifying `[[resolver.nameservers]]` while `resolver.type != "custom"` is currently allowed though.

> it should go in a separate merge request.
Yeah sure, I can separate out editorconfig into a separate MR.

williamdes commented:
I tested this MR in my Docker build and it works fine!

mdecimus commented:
Hi,

Thanks for suggesting this feature. I had to re-implement it a bit differently in order to integrate it with the webadmin:

```toml
[resolver]
type = "custom"
custom = ["udp://192.168.1.1:53", "tcp://192.168.1.30", "tls://192.168.1.44:55", "10.0.0.1", "10.0.0.2"]
```
